<a href="https://colab.research.google.com/github/Barikko/Yandex.Practicum-Projects/blob/main/%D0%A4%D0%B8%D0%BD%D0%B0%D0%BB_SQL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Проект по SQL

## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Описание данных

### Таблица books

Содержит данные о книгах:
    
    book_id — идентификатор книги;
    
    author_id — идентификатор автора;

    title — название книги;

    num_pages — количество страниц;

    publication_date — дата публикации книги;

    publisher_id — идентификатор издателя.

In [2]:
query='''
select * from books limit 10
    '''

pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

### Таблица authors

Содержит данные об авторах:

    author_id — идентификатор автора;
    
    author — имя автора.

In [3]:
query='''
select * from authors limit 10
    '''

pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

### Таблица publishers

Содержит данные об издательствах:

    publisher_id — идентификатор издательства;

    publisher — название издательства;

In [4]:
query='''
select * from publishers limit 10
    '''

pd.io.sql.read_sql(query, con = engine) 

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

### Таблица ratings

Содержит данные о пользовательских оценках книг:

    rating_id — идентификатор оценки;

    book_id — идентификатор книги;

    username — имя пользователя, оставившего оценку;

    rating — оценка книги.

In [5]:
query='''
select * from ratings limit 10
    '''

pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

### Таблица reviews

Содержит данные о пользовательских обзорах:

    review_id — идентификатор обзора;

    book_id — идентификатор книги;

    username — имя автора обзора;

    text — текст обзора.

In [6]:
query='''
select * from reviews limit 10
    '''

pd.io.sql.read_sql(query, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## Цели исследования

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Исследование

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [7]:
query='''
select count(book_id) from books
where publication_date > '1/1/2000'
limit 20
'''
pd.io.sql.read_sql(query, con = engine)

count
0    819

После 1 января 2000 года было выпущено 819 книг

###  Для каждой книги посчитайте количество обзоров и среднюю оценку

In [8]:
query='''
select title, book_id, avg(ratings.rating), count(distinct reviews.review_id) from books
left join ratings using (book_id)
left join reviews using (book_id)
group by books.book_id
order by count(distinct reviews.review_id)  desc
'''
pd.io.sql.read_sql(query, con = engine)

title  book_id       avg  \
0                              Twilight (Twilight  #1)      948  3.662500   
1                                  Water for Elephants      963  3.977273   
2                                     The Glass Castle      734  4.206897   
3    Harry Potter and the Prisoner of Azkaban (Harr...      302  4.414634   
4    The Curious Incident of the Dog in the Night-Time      695  4.081081   
..                                                 ...      ...       ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel       83  3.666667   
996                            The Natural Way to Draw      808  3.000000   
997   The Cat in the Hat and Other Dr. Seuss Favorites      672  5.000000   
998                          Essential Tales and Poems      221  4.000000   
999  Disney's Beauty and the Beast (A Little Golden...      191  4.000000   

     count  
0        7  
1        6  
2        6  
3        6  
4        6  
..     ...  
995      0  
996      0  
997      0  
998      0  
999      0  

[1000 rows x 4 columns]

Наибольшее количество обзоров имеет книга Twilight (7), большинство оценок колеблются от 3.5 до 4.5 

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query='''

select count(book_id) c, publisher  from books
join publishers using (publisher_id)
where num_pages>50
group by publisher
order by c desc
limit 1
'''

pd.io.sql.read_sql(query, con = engine)

c      publisher
0  42  Penguin Books

Наибольшее число книг выпустило издательство Penguin Books - 42.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [10]:
query='''
select author,
       round(avg(rating), 2) as avg_rating
from authors as a
join books as b on b.author_id=a.author_id
join ratings as r on r.book_id=b.book_id
where b.book_id in (select  book_id
                            from ratings
                            group by book_id
                            having count(rating) >= 50)
group by author
order by avg(rating) desc
limit 1

'''
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

J.K. Rowling  имеет самую высокую среднюю оценку - 4.29.

### Посчитайте среднее количество обзоров от пользователей, которые поставили более чем по 50 оценок

In [11]:
query = '''
select round(avg(count)) 
from (select count(text)
from reviews
join (select username, count(rating_id) as cnt
from ratings
group by username) as cnt_rating
on cnt_rating.username = reviews.username
where cnt > 50
group by reviews.username) as count;                                  
'''


pd.io.sql.read_sql(query, con = engine)

round
0   24.0

Среднее количество обзоров - 24